# Imports

In [1]:
# Importing required libraries for the project
import sys # for python library version
import numpy as np # for scientific computing
import pandas as pd # for data anaysis
import matplotlib # for visualization
import seaborn as sns # for visualization
import sklearn # ML Library

# Tuesday DF

In [2]:
print('Python: {}'.format(sys.version))  # Python version
print('numpy: {}'.format(np.__version__))  # Numpy version
print('pandas: {}'.format(pd.__version__))  # Pandas version
print('matplotlib: {}'.format(matplotlib.__version__))  # Matplotlib version
print('seaborn: {}'.format(sns.__version__))  # seaborn version
print('sklearn: {}'.format(sklearn.__version__))  # sklearn version

Python: 3.6.6 |Anaconda, Inc.| (default, Jun 28 2018, 11:27:44) [MSC v.1900 64 bit (AMD64)]
numpy: 1.15.2
pandas: 0.23.4
matplotlib: 3.0.1
seaborn: 0.8.1
sklearn: 0.20.0


In [3]:
# No warning of any kind please!
import warnings
# will ignore any warnings
warnings.filterwarnings("ignore")

# Getting Started

First thing first, we need to import/read the dataset and have a peak at it....

In [4]:
# importing the dataset to a variable
data = pd.read_csv("K:/CIC-2017-dataset/CIC-IDS-2017/MachineLearningCVE/Tuesday-WorkingHours.pcap_ISCX.csv")

# displaying first 3 observations
data.head(3)

Destination Port   Flow Duration   Total Fwd Packets  \
0                 88             640                   7   
1                 88             900                   9   
2                 88            1205                   7   

    Total Backward Packets  Total Length of Fwd Packets  \
0                        4                          440   
1                        4                          600   
2                        4                         2776   

    Total Length of Bwd Packets   Fwd Packet Length Max  \
0                           358                     220   
1                          2944                     300   
2                          2830                    1388   

    Fwd Packet Length Min   Fwd Packet Length Mean   Fwd Packet Length Std  \
0                       0                62.857143              107.349008   
1                       0                66.666667              132.287566   
2                       0               396.571429              677.274651   

    ...     min_seg_size_forward  Active Mean   Active Std   Active Max  \
0   ...                       20          0.0          0.0            0   
1   ...                       20          0.0          0.0            0   
2   ...                       20          0.0          0.0            0   

   Active Min Idle Mean   Idle Std   Idle Max   Idle Min   Label  
0           0       0.0        0.0          0          0  BENIGN  
1           0       0.0        0.0          0          0  BENIGN  
2           0       0.0        0.0          0          0  BENIGN  

[3 rows x 79 columns]

We can see that the data has been imported successfully. Now we need to know the number of observations and features we have.

In [5]:
# dimensions of the data
# where x will be no. of observation
# and y will be features including 1 target variable
x, y = data.shape   # x=445909   y=79

print('We have ', x, ' number of observations and ', y-1, ' features for this dataset to predict type of traffic.')  # removing count of a target variable in 'y'

We have  445909  number of observations and  78  features for this dataset to predict type of traffic.


Let's look at the datatypes of each feature and see if it needs any processing if the feature is not in its appropriate form

In [6]:
# datatypes of features
data.dtypes

 Destination Port                 int64
 Flow Duration                    int64
 Total Fwd Packets                int64
 Total Backward Packets           int64
Total Length of Fwd Packets       int64
 Total Length of Bwd Packets      int64
 Fwd Packet Length Max            int64
 Fwd Packet Length Min            int64
 Fwd Packet Length Mean         float64
 Fwd Packet Length Std          float64
Bwd Packet Length Max             int64
 Bwd Packet Length Min            int64
 Bwd Packet Length Mean         float64
 Bwd Packet Length Std          float64
Flow Bytes/s                     object
 Flow Packets/s                  object
 Flow IAT Mean                  float64
 Flow IAT Std                   float64
 Flow IAT Max                     int64
 Flow IAT Min                     int64
Fwd IAT Total                     int64
 Fwd IAT Mean                   float64
 Fwd IAT Std                    float64
 Fwd IAT Max                      int64
 Fwd IAT Min                      int64


In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 445909 entries, 0 to 445908
Data columns (total 79 columns):
 Destination Port               445909 non-null int64
 Flow Duration                  445909 non-null int64
 Total Fwd Packets              445909 non-null int64
 Total Backward Packets         445909 non-null int64
Total Length of Fwd Packets     445909 non-null int64
 Total Length of Bwd Packets    445909 non-null int64
 Fwd Packet Length Max          445909 non-null int64
 Fwd Packet Length Min          445909 non-null int64
 Fwd Packet Length Mean         445909 non-null float64
 Fwd Packet Length Std          445909 non-null float64
Bwd Packet Length Max           445909 non-null int64
 Bwd Packet Length Min          445909 non-null int64
 Bwd Packet Length Mean         445909 non-null float64
 Bwd Packet Length Std          445909 non-null float64
Flow Bytes/s                    445708 non-null object
 Flow Packets/s                 445909 non-null object
 Flow IAT Mean 

# Explanation & Exploration of the data

Our dataset has 78 features and 1 target variable ' Label'. From 78 features, 60 are numeric and 18 are catrgorical. From 18 categorical, 10 are the features that always contain only one value 0. So we can delete that kind of redundant catrgorical columns whose values are always zero.

Test if there any missing values in DataFrame. It turns out there are missing values

In [8]:
data.isnull().sum()

 Destination Port                 0
 Flow Duration                    0
 Total Fwd Packets                0
 Total Backward Packets           0
Total Length of Fwd Packets       0
 Total Length of Bwd Packets      0
 Fwd Packet Length Max            0
 Fwd Packet Length Min            0
 Fwd Packet Length Mean           0
 Fwd Packet Length Std            0
Bwd Packet Length Max             0
 Bwd Packet Length Min            0
 Bwd Packet Length Mean           0
 Bwd Packet Length Std            0
Flow Bytes/s                    201
 Flow Packets/s                   0
 Flow IAT Mean                    0
 Flow IAT Std                     0
 Flow IAT Max                     0
 Flow IAT Min                     0
Fwd IAT Total                     0
 Fwd IAT Mean                     0
 Fwd IAT Std                      0
 Fwd IAT Max                      0
 Fwd IAT Min                      0
Bwd IAT Total                     0
 Bwd IAT Mean                     0
 Bwd IAT Std                

In [9]:
data.isnull().values.any()

True

Removing Observation which has any Missing Values in it....

In [10]:
# will delete observation if it has any missing values in any of the features.
data.dropna()

# shape of the data after deleting missing entries
data.shape

(445909, 79)

NO Missing Values...!! That's great!

Handling Duplicates

In [11]:
# deleting duplicates, except the first observation
data.drop_duplicates(keep='first')

# shape of the data after deleting duplicate entries
data.shape

(445909, 79)

NO Duplicates too..! Neat!

# Feature statistics --- feature describe

We will split the data in 2 parts. First part will contain all numerical features and second part will contain all binary or categorical features of the data. The target variable Cover_Type is excluded.

In [12]:
# Extracting all numerical features from data
num_fea = data[[" Destination Port",
           " Flow Duration",
           " Total Fwd Packets",
           " Total Backward Packets",
           "Total Length of Fwd Packets",
           " Total Length of Bwd Packets",
           " Down/Up Ratio",
           " Fwd Packet Length Max",
           " Fwd Packet Length Min",
           " Fwd Packet Length Mean",
           " Fwd Packet Length Std",
           "Bwd Packet Length Max",
           " Bwd Packet Length Min",
           " Bwd Packet Length Mean",
           " Bwd Packet Length Std",
           "Flow Bytes/s",
           " Flow Packets/s",
           " Flow IAT Mean",
           " Flow IAT Std",
           " Flow IAT Max",
           " Flow IAT Min", 
           "Fwd IAT Total",
           " Fwd IAT Mean",
           " Fwd IAT Std",
           " Fwd IAT Max", 
           " Fwd IAT Min",
           "Bwd IAT Total", 
           " Bwd IAT Mean",
           " Bwd IAT Std", 
           " Bwd IAT Max", 
           " Bwd IAT Min", 
           " Fwd Header Length",
           " Bwd Header Length",
           "Fwd Packets/s",
           " Bwd Packets/s", 
           " Min Packet Length",
           " Max Packet Length",
           " Packet Length Mean",
           " Packet Length Std",
           " Packet Length Variance",
           " Avg Bwd Segment Size",
           " Average Packet Size",
           " Avg Fwd Segment Size",
           " Fwd Header Length",
           "Subflow Fwd Packets", 
           " Subflow Fwd Bytes", 
           " Subflow Bwd Packets",
           " Subflow Bwd Bytes",
           "Init_Win_bytes_forward", 
           " Init_Win_bytes_backward", 
           " act_data_pkt_fwd",
           " min_seg_size_forward",
           "Active Mean",
           " Active Std",
           " Active Max",
           " Active Min",
           "Idle Mean",
           " Idle Std",
           " Idle Max",
           " Idle Min"

 ]]

# extracting all binary/ categorical features from data
binary_fea = data[["Fwd PSH Flags",
           " Bwd PSH Flags",  # 0
           " Fwd URG Flags",  # 0
           " Bwd URG Flags",  # 0
           "FIN Flag Count", 
           " SYN Flag Count",
           " RST Flag Count",
           " PSH Flag Count",
           " ACK Flag Count",
           " URG Flag Count",
           " CWE Flag Count",  # 0
           " ECE Flag Count",
           "Fwd Avg Bytes/Bulk",  # 0
           " Fwd Avg Packets/Bulk", # 0
           " Fwd Avg Bulk Rate",  # 0
           " Bwd Avg Bytes/Bulk",  # 0
           " Bwd Avg Packets/Bulk", # 0
           "Bwd Avg Bulk Rate"]]


In [13]:
num_fea.shape     #  (445909,60)

(445909, 60)

In [14]:
num_fea.describe()

Destination Port   Flow Duration   Total Fwd Packets  \
count      445909.000000    4.459090e+05       445909.000000   
mean         8878.375745    1.077944e+07           11.800473   
std         19522.889728    2.955629e+07          867.072428   
min             0.000000   -4.000000e+00            1.000000   
25%            53.000000    1.860000e+02            2.000000   
50%            80.000000    3.128800e+04            2.000000   
75%           443.000000    4.668200e+05            4.000000   
max         65505.000000    1.200000e+08       206446.000000   

        Total Backward Packets  Total Length of Fwd Packets  \
count            445909.000000                 4.459090e+05   
mean                 13.907212                 5.302105e+02   
std                1172.161547                 5.676904e+03   
min                   0.000000                 0.000000e+00   
25%                   1.000000                 4.000000e+00   
50%                   2.000000                 6.800000e+01   
75%                   2.000000                 1.500000e+02   
max              276072.000000                 2.428415e+06   

        Total Length of Bwd Packets   Down/Up Ratio   Fwd Packet Length Max  \
count                  4.459090e+05    445909.00000           445909.000000   
mean                   2.186324e+04         0.68745              173.346669   
std                    2.625768e+06         0.53444              472.450792   
min                    0.000000e+00         0.00000                0.000000   
25%                    0.000000e+00         0.00000                1.000000   
50%                    1.360000e+02         1.00000               41.000000   
75%                    3.360000e+02         1.00000               62.000000   
max                    6.270000e+08        35.00000            24820.000000   

        Fwd Packet Length Min   Fwd Packet Length Mean      ...       \
count           445909.000000            445909.000000      ...        
mean                20.579979                48.528489      ...        
std                 38.938350               113.195021      ...        
min                  0.000000                 0.000000      ...        
25%                  0.000000                 0.857143      ...        
50%                  0.000000                37.000000      ...        
75%                 41.000000                50.000000      ...        
max               2065.000000              4672.000000      ...        

        act_data_pkt_fwd   min_seg_size_forward   Active Mean    Active Std  \
count      445909.000000           4.459090e+05  4.459090e+05  4.459090e+05   
mean            2.088341          -1.321786e+04  7.265964e+04  4.670517e+04   
std             6.325197           2.666478e+06  6.205625e+05  3.601504e+05   
min             0.000000          -5.368707e+08  0.000000e+00  0.000000e+00   
25%             0.000000           2.000000e+01  0.000000e+00  0.000000e+00   
50%             1.000000           2.000000e+01  0.000000e+00  0.000000e+00   
75%             2.000000           3.200000e+01  0.000000e+00  0.000000e+00   
max          1846.000000           1.380000e+02  1.070000e+08  2.410000e+07   

         Active Max    Active Min     Idle Mean      Idle Std      Idle Max  \
count  4.459090e+05  4.459090e+05  4.459090e+05  4.459090e+05  4.459090e+05   
mean   1.642092e+05  4.828426e+04  3.294459e+06  1.497105e+05  3.407025e+06   
std    1.049887e+06  5.649489e+05  1.258776e+07  1.982346e+06  1.293249e+07   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
50%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
75%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
max    1.070000e+08  1.070000e+08  1.200000e+08  7.590000e+07  1.200000e+08   

           Idle Min  
count  4.459090e+05  
mean   3.145407e+06  
std    1.240

In [15]:
binary_fea.shape

(445909, 18)

In [16]:
listLabel = list(binary_fea["Bwd Avg Bulk Rate"].drop_duplicates().values)
print('\n The types of all traffics are {}'.format(listLabel))


 The types of all traffics are [0]


So we can delete the redundant feature columns whose values are always zero

In [17]:
binary_fea = binary_fea.drop([" Bwd PSH Flags",  # 0
           " Fwd URG Flags",  # 0
           " Bwd URG Flags",
           " CWE Flag Count",
           "Fwd Avg Bytes/Bulk",  # 0
           " Fwd Avg Packets/Bulk", # 0
           " Fwd Avg Bulk Rate",  # 0
           " Bwd Avg Bytes/Bulk",  # 0
           " Bwd Avg Packets/Bulk", # 0
           "Bwd Avg Bulk Rate"], axis=1)

In [18]:
binary_fea.describe()

Fwd PSH Flags  FIN Flag Count   SYN Flag Count   RST Flag Count  \
count  445909.000000   445909.000000    445909.000000    445909.000000   
mean        0.060723        0.018380         0.060723         0.000336   
std         0.238822        0.134323         0.238822         0.018338   
min         0.000000        0.000000         0.000000         0.000000   
25%         0.000000        0.000000         0.000000         0.000000   
50%         0.000000        0.000000         0.000000         0.000000   
75%         0.000000        0.000000         0.000000         0.000000   
max         1.000000        1.000000         1.000000         1.000000   

        PSH Flag Count   ACK Flag Count   URG Flag Count   ECE Flag Count  
count    445909.000000    445909.000000    445909.000000    445909.000000  
mean          0.242597         0.291687         0.115091         0.000339  
std           0.428654         0.454540         0.319132         0.018399  
min           0.000000         0.000000         0.000000         0.000000  
25%           0.000000         0.000000         0.000000         0.000000  
50%           0.000000         0.000000         0.000000         0.000000  
75%           0.000000         1.000000         0.000000         0.000000  
max           1.000000         1.000000         1.000000         1.000000

Now that we have our separate numerical DataFrame 'num_fea', it's time to feature-scale it

In [19]:
num_fea = num_fea.rename(columns={"Flow Bytes/s":"Flow Bytes", " Flow Packets/s":"Flow Packets", "Fwd Packets/s":"Fwd Packets", " Bwd Packets/s":"Bwd Packets"})


before we move to splitting our data to Train-Test Split is to scale the features to some specific range. but before we do this we will split the feature and target variables because we dont want to scale our target variable..

In [20]:
num_fea = num_fea.drop(['Flow Bytes','Flow Packets'], axis=1)

Now we can re-merge our two DataFrames into a new, X_scaled 

# Class Distribution:
Let's take a look how each class is distributed..

We have uneven samples of traffic type, where BENIGN has the highest no. of observation.
But we do have enough samples to train the model learning different patterns of each traffic types. We will see how models performs with these uneven amount of distributions in Model Evaluation section.

In [37]:
# grouping by forest cover type and calculating total occurance
data.groupby(' Label').size()

 Label
BENIGN         432074
FTP-Patator      7938
SSH-Patator      5897
dtype: int64

In [38]:
data.loc[data[' Label'] == 'BENIGN'] = 0
data.loc[data[' Label'] == 'FTP-Patator'] = 1
data.loc[data[' Label'] == 'SSH-Patator'] = 2

Now we can re-merge our two DataFrames into a new, X_scaled 

In [39]:
X = pd.concat([binary_fea, num_fea], axis=1, join='inner')    # 445909 × 66
# feeding our target variable to var 'y'
y = data[' Label']

In [42]:
# importing feature scaling function
from sklearn.preprocessing import MinMaxScaler

# passing range to the function and then save it
scaler = MinMaxScaler(feature_range = (0,1))


# apply feature scaling to all features
X_scaled = scaler.fit_transform(X)

#s_sample_2 = scaler.fit_transform(X2)

# Train-Test Split
Now our data is ready to be splitted into 75%-25% train-test set respectively

In [43]:
# importing train-test function
from sklearn.model_selection import train_test_split

# split the data in 75%-25% train-test respectively with fixed state
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size = 0.25, random_state = 53)

In [44]:
# number of training observation
print(X_train.shape, X_test.shape)

(334431, 66) (111478, 66)


Lots of data to train and test on.

# Model Evaluations

Now its time to feed our data to the models to see how each models performs using 2 different evaluation metrics accuracy and f1 score and see which model performs the best.

But before that, we will train our data on training set and test the performance of the Benchmark model we discussed about in the start of the project. I will use 10 K-Fold CV to test the performance of our model. I had choosen Naive Bayes Classifier as my benchmark model and I am going to use Multimonial Naive Bayes classifier since we have a claasification problem to solve.

The Evaluation Metric I am going to use are f1 score and accuracy to see how well our model performs.

Accuracy is the measure of the correct predicted data divided by total number of observations hence giving a value ranging between 0 and 1, while 0 is no correctly predicted class whereas 1 is all correctly predicted class. We can multiply the result by 100 to get the accuracy score in terms of percent.

F1 score is more useful than accuracy specially in the case where you have uneven amount of class distribution as in our case. It's the weighted average of Precision and Recall. Therefore, this score takes both false positives and false negatives into account.

Accuracy works best if false positives and false negatives have similar cost. If the cost of false positives and false negatives are very different, it’s better to look at both Precision and Recall or F1 score.

First I will define a function which will train the models using training data and calculate model's performance using accuracy and f1 score. One sets of instruction for all models!

In [45]:
### defining function for training models and measuring performance 

# to measure performance
from sklearn.model_selection import cross_val_score

# for calculating time elapsed
import time

# fucntion
def model_evaluation(clf):
    
    # passing classifier to a variable
    clf = clf
    
    # records time
    t_start = time.time()
    # classifier learning the model
    clf = clf.fit(X_train, y_train)
    # records time
    t_end = time.time()
    
    
    # records time
    c_start = time.time()     
    # Using 10 K-Fold CV on data, gives peroformance measures
    accuracy  = cross_val_score(clf, X_train, y_train, cv = 10, scoring = 'accuracy')
    f1_score = cross_val_score(clf, X_train, y_train, cv = 10, scoring = 'f1_macro')
    # records the time
    c_end = time.time()    
    
    
    # calculating mean of all 10 observation's accuracy and f1, taking percent and rounding to two decimal places
    acc_mean = np.round(accuracy.mean() * 100, 2)
    f1_mean = np.round(f1_score.mean() * 100, 2)
    
    # substracts end time with start to give actual time taken in seconds
    # divides by 60 to convert in minutes and rounds the answer to three decimal places
    # time in training
    t_time = np.round((t_end - t_start) / 60, 3)
    # time for evaluating scores
    c_time = np.round((c_end - c_start) / 60, 3)
    
    
    # Removing traces of classifier
    clf = None
    
    
    # returns performance measure and time of the classifier 
    print("The accuracy score of this classifier on our training set is", acc_mean,"% and f1 score is", f1_mean,"% taking", t_time,"minutes to train and", c_time,
          "minutes to evaluate cross validation and metric scores.")


# Benchmark Mode

Now lets see the performance of MultinomialNB classifier on given training data.

In [46]:
# importing Multinomial Naive Bayes classifier, one of the Naive Bayes classifier
from sklearn.naive_bayes import MultinomialNB

# passing the model to function to get performance measures
model_evaluation(MultinomialNB())

The accuracy score of this classifier on our training set is 96.79 % and f1 score is 49.03 % taking 0.006 minutes to train and 0.14 minutes to evaluate cross validation and metric scores.


In [47]:
# importing K-Nearest Neighbors Classifier function
from sklearn.neighbors import KNeighborsClassifier

model_evaluation(KNeighborsClassifier(n_jobs=-1))

The accuracy score of this classifier on our training set is 99.97 % and f1 score is 99.65 % taking 8.777 minutes to train and 184.721 minutes to evaluate cross validation and metric scores.


In [48]:
# importing Random Forest function
from sklearn.ensemble import RandomForestClassifier

model_evaluation(RandomForestClassifier(n_jobs=-1, random_state = 53))

The accuracy score of this classifier on our training set is 100.0 % and f1 score is 99.97 % taking 0.037 minutes to train and 0.763 minutes to evaluate cross validation and metric scores.


In [49]:
# importing Stochastic Gradient Descent Classifier function
from sklearn.linear_model import SGDClassifier

model_evaluation(SGDClassifier(n_jobs=-1, random_state = 53))

The accuracy score of this classifier on our training set is 96.87 % and f1 score is 32.8 % taking 0.033 minutes to train and 0.335 minutes to evaluate cross validation and metric scores.


In [50]:
# importing AdaBoost classifier
from sklearn.ensemble import ExtraTreesClassifier

model_evaluation(ExtraTreesClassifier(n_jobs=-1, random_state = 53))

The accuracy score of this classifier on our training set is 99.99 % and f1 score is 99.85 % taking 0.064 minutes to train and 1.232 minutes to evaluate cross validation and metric scores.


In [51]:
from sklearn.linear_model import LogisticRegression

model_evaluation(LogisticRegression(n_jobs = -1, random_state = 53))

The accuracy score of this classifier on our training set is 98.17 % and f1 score is 62.35 % taking 0.354 minutes to train and 6.076 minutes to evaluate cross validation and metric scores.


# Choosing Model
Out of 5 Models evaluated above and benchmark model, which performs better? Lets see all the scores of all the models in a table below:

RF performs here stunningly awesome, getting accuracy 100% and 99.97% of F1 score. taking seconds to train and couple of minutes running cross validation and metrics results. Also given its flexibility it has perofrmed so well with default params.

# Testing Model

Since I had chosen default params above in Random Forest in training set, I will be tuning important param that might give us more better results. Later I will discuss about it.

In [52]:
# importing EM scores for model performance measure
from sklearn.metrics import accuracy_score, f1_score

# definning best chosen classifier
clf = RandomForestClassifier(n_estimators = 50, random_state = 53)

# training our model
clf = clf.fit(X_train, y_train)

# predicting unseen data
predict = clf.predict(X_test)

# calculating accuracy
accuracy = accuracy_score(y_test, predict)

# calculating f1 score
f1_score = f1_score(y_test, predict, average = 'macro')

# taking precentage and rounding to 3 places
accuracy = np.round(accuracy * 100, 3)
f1_score = np.round(f1_score * 100, 3)

# cleaning traces
clf = None

# results
print("The accuracy score of our final model Random Forest Classifier on our testing set is", accuracy,"% and f1 score is", f1_score,"%.")

The accuracy score of our final model Random Forest Classifier on our testing set is 99.995 % and f1 score is 99.933 %.
